pyrealm/pmodel/calc_optimal_chi_new.py

```diff
@@ -0,0 +1,573 @@
+      :math:`\beta`).
+    - The variable ``xi`` (:math:`\xi`), which captures the sensitivity of :math:`\chi`
+      to the vapour pressure deficit, and is related to the carbon cost of water (Medlyn
+      et al. 2011; Prentice et 2014).
+    - The optimal ratio of :math:`\ce{CO2}` partial pressures within the leaf (``ci``,
+      :math:`c_i`) to the external environmental partial pressure (``ca```, :math:`c_i`)
+      recorded. The optimal ratio ``chi`` (:math:`\chi = c_i/c_a`) is stored as well as
+      the resulting internal partial pressure (``ci``).
+    - The :math:`\ce{CO2}` limitation term for light-limited assimilation (``mj``,
+      :math:`m_j`), the :math:`\ce{CO2}` limitation term for Rubisco-limited
+      assimilation (``mc``, :math:`m_c`) and their ratio (``mjoc``, :math:`m_j/m_c`).
+
+    The abstract base class requires that implementations of specific approaches define
+    two methods:
+
+    - `set_beta`: This method defines the calculation of the ``beta`` values to be used.
+      The method is called by the ``__init__`` method when a subclass instance is
+      created and should not change.
+    - `estimate_chi`: This method defines the calculation of ``xi`` and then ``chi``,
+      ``ci``, ``mj``, ``mc`` and ``mjoc``. This method is also called by the
+      ``__init__`` method when a subclass instance is created, using the default defined
+      calculation of ``xi``. However, it can also be called using modified values for
+      ``xi`` to allow recalculation of these values. This is used primarily in fitting
+      the P Model at subdaily scales, where the ``xi`` parameter acclimates slowly to
+      changing environmental conditions.
+
+    Args:
+        env: An instance of
+            :class:`~pyrealm.pmodel.pmodel_environment.PModelEnvironment`  providing the
+            photosynthetic environment for the model.
+        rootzonestress: This is an experimental feature to supply a root zone stress
+            factor used as a direct penalty to :math:`\beta`, unitless. The default is
+            1.0, with no root zone stress applied.
+        pmodel_const: An instance of
+            :class:`~pyrealm.constants.pmodel_const.PModelConst`.
+
+    Returns:
+        Instances of the abstract base class should not be created - use instances of
+        specific subclasses.
+    """
+
+    # @property
+    # @abstractmethod
+    # def method(cls) -> str:
+    #     """The method name.
+
+    #     This class property sets the name used to refer to identify the class in
+    #     the :data:`~pyrealm.pmodel.calc_optimal_chi.OPTIMAL_CHI_METHOD_REGISTRY`.
+    #     """
+
+    # @property
+    # @abstractmethod
+    # def is_c4(cls) -> bool:
+    #     """Does the class represent a C4 pathway."""
+
+    method: str
+    is_c4: bool
+
+    def __init__(
+        self,
+        env: PModelEnvironment,
+        rootzonestress: NDArray = np.array([1.0]),
```

Comment:
The default value of `rootzonestress` is mutable, so if there is an instance `opt_chi` and a user sets `opt_chi.rootzonestress[0] = 2.0`, the default value of `rootzonestress` of all instances of `NewCalcOptimalChi` will also be changed. This is probably not desirable. See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.

Reply:
Good point. I guess the canonical way of handling this is to do `rootzonestress: None | NDArray = None` and then populate it in `__init__`.

I was thinking of pushing `rootzonestress` back into `PModelEnvironment` - `theta` there is handled in this way and it would be more consistent to pass all the variables via that object. I can then have `rootzonestress` specific subclasses of `CalcOptimalChiNew`.

Reply:
That doesn't conform to the `NDArray` typing though

Reply:
Yes, but that's not obvious because the `SplashModel` object is missing typing on the attributes - those should all match the typing from the `__init__` signature.